pkg/konnector/controllers/cluster/serviceexport/cluster-scoped/utils.go

```diff
@@ -0,0 +1,123 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterscoped
+
+import (
+	"errors"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func Prepend(name, clusterNs string) string {
+	return clusterNs + "-" + name
+}
+
+func Behead(name, clusterNs string) string {
+	return strings.TrimPrefix(name, clusterNs+"-")
+}
+
+// SwitchToUpstreamName switches the name of a downstream
+// cluster-scoped object by prepending the cluster namespace name
+func SwitchToUpstreamName(obj *unstructured.Unstructured, clusterNs string) {
+	downstreamName := obj.GetName()
+	upstreamName := Prepend(downstreamName, clusterNs)
+	obj.SetName(upstreamName)
+}
+
+// SwitchToDownstreamName switches the name of a upstream
+// cluster-scoped object by removing the cluster namespace name as the prefix
+func SwitchToDownstreamName(obj *unstructured.Unstructured, clusterNs string) {
+	upstreamName := obj.GetName()
+	downstreamName := Behead(upstreamName, clusterNs)
+	obj.SetName(downstreamName)
+}
+
+// InjectClusterNs injects the corresponding cluster namespace, by design as the
+// single owner reference, into a cluster-scoped object.
+func InjectClusterNs(obj *unstructured.Unstructured, clusterNs, clusterNsUID string) error {
+	ors := obj.GetOwnerReferences()
+	if len(ors) != 0 {
+		return errors.New("object already owned by something else")
+	}
+	ors = append(ors, metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Namespace",
+		Name:       clusterNs,
+		UID:        types.UID(clusterNsUID),
+	})
+	obj.SetOwnerReferences(ors)
+	return nil
+}
+
+// ExtractClusterNs extracts the corresponding cluster namespace name
+// from a cluster-scoped object.
+func ExtractClusterNs(obj *unstructured.Unstructured) (string, error) {
+	ors := obj.GetOwnerReferences()
+	if len(ors) == 0 {
+		return "", errors.New("owner references not found in object")
+	}
+	if len(ors) > 1 {
+		return "", errors.New("object has multiple owners")
+	}
+	if ors[0].Kind != "Namespace" {
+		return "", errors.New("object owned by something other than a namespace")
+	}
+	return ors[0].Name, nil
+}
+
+// ClearClusterNs clears the corresponding cluster namespace
+// in a cluster-scoped object.
+func ClearClusterNs(obj *unstructured.Unstructured) error {
+	ors := obj.GetOwnerReferences()
+	if len(ors) == 0 {
+		return errors.New("owner references not found in object")
+	}
+	if len(ors) > 1 {
+		return errors.New("object has multiple owners")
+	}
+	if ors[0].Kind != "Namespace" {
+		return errors.New("object owned by something other than a namespace")
+	}
+	ors = ors[:0]
+	obj.SetOwnerReferences(ors)
+	return nil
+}
+
+func TranslateFromDownstream(obj *unstructured.Unstructured, clusterNs, clusterNsUID string) (*unstructured.Unstructured, error) {
```

Comment:
this mutates. Intentional? The signature does not look like that. Would either rename to SomethingInPlace or better remove the unstructured return value. Same below.

Reply:
Thank you. Agreed and removed the unstructured return values in 94cf53e.

Reply:
in a later commit? Please squash the later change here.